rust/cubesql/cubesql/src/compile/engine/udf/common.rs

```diff
@@ -484,83 +484,215 @@ pub fn create_if_udf() -> ScalarUDF {
 
-        if types[0] == DataType::Null || types[1] == DataType::Null {
-            return Ok(Arc::new(DataType::Null));
-        }
+        let mut ti = types.iter();
+        let mut base_type = ti.next().unwrap().clone();
 
-        let base_type = common_type_coercion(&types[0], &types[1]).ok_or_else(|| {
-            DataFusionError::Execution(format!(
-                "Unable to coercion types, actual: [{}, {}]",
-                &types[0], &types[1],
-            ))
-        })?;
+        for t in ti {
+            base_type = common_type_coercion(&base_type, t).ok_or_else(|| {
+                DataFusionError::Execution(format!(
+                    "Unable to coercion types, actual: [{}, {}]",
+                    &base_type, t,
+                ))
+            })?;
+        }
 
         Ok(Arc::new(base_type))
     });
 
     ScalarUDF::new(
         "least",
-        &Signature::any(2, Volatility::Immutable),
+        &Signature::variadic(
+            vec![DataType::Int64, DataType::UInt64, DataType::Float64],
+            Volatility::Immutable,
+        ),
+        &return_type,
+        &fun,
+    )
+}
+
+/// The GREATEST function selects the largest value from a list of any number of expressions.
+/// The expressions must all be convertible to a common data type, which will be the type of
+/// the result. The SQL standard requires greatest and least to return null in case one argument
+/// is null. However, in Postgres NULL values in the list are ignored.
+/// We follow SQL Standard: The result will be NULL if any expression evaluates to NULL.
+pub fn create_greatest_udf() -> ScalarUDF {
+    let fun = make_scalar_function(move |args: &[ArrayRef]| {
+        assert!(args.len() > 0);
+
+        let mut base_type = DataType::Null;
+
+        for arg in args {
+            base_type = common_type_coercion(&base_type, arg.data_type()).ok_or_else(|| {
+                DataFusionError::Execution(format!(
+                    "Unable to coercion types, actual: [{}, {}]",
+                    &base_type,
+                    arg.data_type(),
+                ))
+            })?;
+        }
+
+        // If there is Null value in the list - return Null
+        if let Some(v) = args.iter().find(|v| v.is_null(0)) {
+            return Ok(v.clone());
+        }
+
+        let mut max_value: Option<ArrayRef> = None;
+
+        for arg in args {
+            let casted_arg = cast(&arg, &base_type)?;
+            max_value = match max_value {
+                Some(current_max) => {
+                    let is_greater = match base_type {
+                        DataType::UInt64 => {
+                            let current_max =
+                                downcast_primitive_arg!(current_max, "current_max", UInt64Type);
+                            let casted_arg =
+                                downcast_primitive_arg!(casted_arg, "casted_arg", UInt64Type);
+
+                            current_max.value(0) < casted_arg.value(0)
```

Comment:
This function will not work with tables. You can test it with something like

```sql
SELECT least(t.a, t.b, t.c) FROM (
    SELECT 1 as a, 2 as b, 3 as c UNION ALL
    SELECT 3, 2, 1 UNION ALL
) as t
```

Reply:
Done!